File: 54remitflow-unified-platform/backend/go-services/workflow-orchestrator/internal/integration/payment_processor.go
package integration

import (
	"context"
	"fmt"
	"time"

	"workflow-orchestrator/internal/middleware/redis"
	"workflow-orchestrator/internal/middleware/tigerbeetle"
	"workflow-orchestrator/internal/middleware/kafka"
	"workflow-orchestrator/pkg/logger"
)

// PaymentRequest represents a payment transaction request
type PaymentRequest struct {
	PaymentID       string
	WorkflowID      string
	FromAccountID   tigerbeetle.uint128
	ToAccountID     tigerbeetle.uint128
	Amount          uint64
	Currency        string
	Description     string
	TenantID        string
	UserID          string
	IdempotencyKey  string
}

// PaymentResult represents the result of a payment transaction
type PaymentResult struct {
	PaymentID       string
	TransferID      tigerbeetle.uint128
	Status          string
	Timestamp       time.Time
	FromAccountID   tigerbeetle.uint128
	ToAccountID     tigerbeetle.uint128
	Amount          uint64
	Error           error
}

// ProcessPaymentWithLocking processes a payment transaction with distributed locking
// This is the complete implementation showing TigerBeetle and Redis interaction
func (m *MiddlewareManager) ProcessPaymentWithLocking(
	ctx context.Context,
	req *PaymentRequest,
) (*PaymentResult, error) {
	logger.Logger.Info(
		"Starting payment processing",
		logger.String("payment_id", req.PaymentID),
		logger.String("workflow_id", req.WorkflowID),
		logger.Uint64("amount", req.Amount),
	)

	// Step 1: Validate payment request
	if err := validatePaymentRequest(req); err != nil {
		logger.Logger.Error("Invalid payment request", logger.Error(err))
		return nil, fmt.Errorf("invalid payment request: %w", err)
	}

	// Step 2: Check for duplicate payment using idempotency key in Redis
	idempotencyKey := fmt.Sprintf("payment:idempotency:%s", req.IdempotencyKey)
	existingResult, err := m.Redis.GetWorkflowState(ctx, idempotencyKey)
	if err == nil && existingResult != nil {
		logger.Logger.Info("Payment already processed (idempotent)", 
			logger.String("payment_id", req.PaymentID))
		return &PaymentResult{
			PaymentID: req.PaymentID,
			Status:    "completed",
			Timestamp: time.Now(),
		}, nil
	}

	// Step 3: Acquire distributed lock to prevent concurrent payment processing
	lockName := fmt.Sprintf("payment:lock:%s", req.PaymentID)
	lockTimeout := 30 * time.Second
	
	logger.Logger.Info("Acquiring distributed lock", logger.String("lock_name", lockName))
	locked, err := m.Redis.AcquireLock(ctx, lockName, int(lockTimeout.Seconds()))
	if err != nil {
		logger.Logger.Error("Failed to acquire lock", logger.Error(err))
		return nil, fmt.Errorf("failed to acquire payment lock: %w", err)
	}
	if !locked {
		logger.Logger.Warn("Payment already being processed", 
			logger.String("payment_id", req.PaymentID))
		return nil, fmt.Errorf("payment %s is already being processed", req.PaymentID)
	}
	logger.Logger.Info("Distributed lock acquired", logger.String("lock_name", lockName))

	// Ensure lock is released even if panic occurs
	defer func() {
		if err := m.Redis.ReleaseLock(ctx, lockName); err != nil {
			logger.Logger.Error("Failed to release lock", logger.Error(err))
		} else {
			logger.Logger.Info("Distributed lock released", logger.String("lock_name", lockName))
		}
	}()

	// Step 4: Cache payment state as "pending" in Redis
	pendingState := map[string]interface{}{
		"payment_id":     req.PaymentID,
		"workflow_id":    req.WorkflowID,
		"status":         "pending",
		"from_account":   req.FromAccountID.String(),
		"to_account":     req.ToAccountID.String(),
		"amount":         req.Amount,
		"currency":       req.Currency,
		"timestamp":      time.Now().Unix(),
	}
	
	stateKey := fmt.Sprintf("payment:state:%s", req.PaymentID)
	if err := m.Redis.CacheWorkflowState(ctx, stateKey, pendingState, 3600); err != nil {
		logger.Logger.Error("Failed to cache payment state", logger.Error(err))
		// Continue processing even if caching fails
	}

	// Step 5: Publish payment.initiated event to Kafka
	initiatedEvent := &kafka.WorkflowEvent{
		EventID:      fmt.Sprintf("evt-%s-initiated", req.PaymentID),
		EventType:    "payment.initiated",
		Timestamp:    time.Now(),
		WorkflowID:   req.WorkflowID,
		WorkflowType: "payment",
		Status:       "pending",
		TenantID:     req.TenantID,
		UserID:       req.UserID,
		Data: map[string]interface{}{
			"payment_id":   req.PaymentID,
			"amount":       req.Amount,
			"currency":     req.Currency,
			"from_account": req.FromAccountID.String(),
			"to_account":   req.ToAccountID.String(),
		},
	}
	
	if err := m.PublishWorkflowEvent(ctx, initiatedEvent); err != nil {
		logger.Logger.Error("Failed to publish payment.initiated event", logger.Error(err))
		// Continue processing even if event publishing fails
	}

	// Step 6: Check account balances (optional pre-validation)
	// This could query TigerBeetle for current balances
	logger.Logger.Info("Validating account balances",
		logger.String("from_account", req.FromAccountID.String()))

	// Step 7: Create transfer in TigerBeetle
	logger.Logger.Info("Creating transfer in TigerBeetle",
		logger.String("payment_id", req.PaymentID),
		logger.Uint64("amount", req.Amount))

	// Generate transfer ID from payment ID
	transferID := tigerbeetle.GenerateTransferID(req.PaymentID)

	// Create the transfer using TigerBeetle
	err = m.TigerBeetle.CreateTransfer(
		ctx,
		transferID,
		req.FromAccountID,
		req.ToAccountID,
		req.Amount,
		1, // ledger ID
		1, // code (payment type)
	)

	// Step 8: Handle transfer result
	result := &PaymentResult{
		PaymentID:     req.PaymentID,
		TransferID:    transferID,
		FromAccountID: req.FromAccountID,
		ToAccountID:   req.ToAccountID,
		Amount:        req.Amount,
		Timestamp:     time.Now(),
	}

	if err != nil {
		// Transfer failed
		logger.Logger.Error("TigerBeetle transfer failed", logger.Error(err))
		result.Status = "failed"
		result.Error = err

		// Update cache with failed status
		failedState := map[string]interface{}{
			"payment_id":  req.PaymentID,
			"status":      "failed",
			"error":       err.Error(),
			"timestamp":   time.Now().Unix(),
		}
		m.Redis.CacheWorkflowState(ctx, stateKey, failedState, 3600)

		// Publish payment.failed event
		failedEvent := &kafka.WorkflowEvent{
			EventID:      fmt.Sprintf("evt-%s-failed", req.PaymentID),
			EventType:    "payment.failed",
			Timestamp:    time.Now(),
			WorkflowID:   req.WorkflowID,
			WorkflowType: "payment",
			Status:       "failed",
			TenantID:     req.TenantID,
			UserID:       req.UserID,
			Data: map[string]interface{}{
				"payment_id": req.PaymentID,
				"amount":     req.Amount,
				"error":      err.Error(),
			},
		}
		m.PublishWorkflowEvent(ctx, failedEvent)

		return result, fmt.Errorf("payment processing failed: %w", err)
	}

	// Transfer succeeded
	logger.Logger.Info("TigerBeetle transfer completed successfully",
		logger.String("payment_id", req.PaymentID),
		logger.String("transfer_id", transferID.String()))
	result.Status = "completed"

	// Step 9: Update cache with completed status
	completedState := map[string]interface{}{
		"payment_id":   req.PaymentID,
		"transfer_id":  transferID.String(),
		"status":       "completed",
		"from_account": req.FromAccountID.String(),
		"to_account":   req.ToAccountID.String(),
		"amount":       req.Amount,
		"currency":     req.Currency,
		"timestamp":    time.Now().Unix(),
	}
	
	if err := m.Redis.CacheWorkflowState(ctx, stateKey, completedState, 3600); err != nil {
		logger.Logger.Error("Failed to cache completed payment state", logger.Error(err))
	}

	// Step 10: Store idempotency key to prevent duplicate processing
	if err := m.Redis.CacheWorkflowState(ctx, idempotencyKey, completedState, 86400); err != nil {
		logger.Logger.Error("Failed to cache idempotency key", logger.Error(err))
	}

	// Step 11: Publish payment.completed event to Kafka
	completedEvent := &kafka.WorkflowEvent{
		EventID:      fmt.Sprintf("evt-%s-completed", req.PaymentID),
		EventType:    "payment.completed",
		Timestamp:    time.Now(),
		WorkflowID:   req.WorkflowID,
		WorkflowType: "payment",
		Status:       "completed",
		TenantID:     req.TenantID,
		UserID:       req.UserID,
		Data: map[string]interface{}{
			"payment_id":   req.PaymentID,
			"transfer_id":  transferID.String(),
			"amount":       req.Amount,
			"currency":     req.Currency,
			"from_account": req.FromAccountID.String(),
			"to_account":   req.ToAccountID.String(),
		},
	}
	
	if err := m.PublishWorkflowEvent(ctx, completedEvent); err != nil {
		logger.Logger.Error("Failed to publish payment.completed event", logger.Error(err))
		// Don't fail the payment if event publishing fails
	}

	logger.Logger.Info("Payment processing completed successfully",
		logger.String("payment_id", req.PaymentID),
		logger.String("status", result.Status))

	return result, nil
}

// ProcessPayment is a simplified version that delegates to ProcessPaymentWithLocking
func (m *MiddlewareManager) ProcessPayment(
	ctx context.Context,
	paymentID string,
	fromAccountID, toAccountID tigerbeetle.uint128,
	amount uint64,
) error {
	req := &PaymentRequest{
		PaymentID:      paymentID,
		WorkflowID:     paymentID,
		FromAccountID:  fromAccountID,
		ToAccountID:    toAccountID,
		Amount:         amount,
		Currency:       "NGN",
		Description:    "Payment transaction",
		IdempotencyKey: paymentID,
	}

	result, err := m.ProcessPaymentWithLocking(ctx, req)
	if err != nil {
		return err
	}

	if result.Status != "completed" {
		return fmt.Errorf("payment failed with status: %s", result.Status)
	}

	return nil
}

// validatePaymentRequest validates the payment request
func validatePaymentRequest(req *PaymentRequest) error {
	if req.PaymentID == "" {
		return fmt.Errorf("payment_id is required")
	}
	if req.Amount == 0 {
		return fmt.Errorf("amount must be greater than 0")
	}
	if req.FromAccountID == req.ToAccountID {
		return fmt.Errorf("from_account and to_account must be different")
	}
	if req.IdempotencyKey == "" {
		return fmt.Errorf("idempotency_key is required")
	}
	return nil
}

// GetPaymentStatus retrieves the current status of a payment from Redis cache
func (m *MiddlewareManager) GetPaymentStatus(
	ctx context.Context,
	paymentID string,
) (map[string]interface{}, error) {
	stateKey := fmt.Sprintf("payment:state:%s", paymentID)
	return m.Redis.GetWorkflowState(ctx, stateKey)
}

// CancelPendingPayment attempts to cancel a pending payment
func (m *MiddlewareManager) CancelPendingPayment(
	ctx context.Context,
	paymentID string,
) error {
	// Acquire lock
	lockName := fmt.Sprintf("payment:lock:%s", paymentID)
	locked, err := m.Redis.AcquireLock(ctx, lockName, 30)
	if err != nil || !locked {
		return fmt.Errorf("failed to acquire lock for cancellation")
	}
	defer m.Redis.ReleaseLock(ctx, lockName)

	// Check current status
	stateKey := fmt.Sprintf("payment:state:%s", paymentID)
	state, err := m.Redis.GetWorkflowState(ctx, stateKey)
	if err != nil {
		return fmt.Errorf("failed to get payment state: %w", err)
	}

	status, ok := state["status"].(string)
	if !ok || status != "pending" {
		return fmt.Errorf("payment cannot be cancelled (status: %s)", status)
	}

	// Update status to cancelled
	state["status"] = "cancelled"
	state["cancelled_at"] = time.Now().Unix()
	
	return m.Redis.CacheWorkflowState(ctx, stateKey, state, 3600)
}

File: 54remitflow-unified-platform/backend/python-services/workflow-orchestrator-enhanced/integration/payment_processor.py
"""
Payment Processor - Complete implementation with TigerBeetle and Redis integration
"""
import logging
import time
from typing import Dict, Any, Optional
from dataclasses import dataclass
from datetime import datetime

logger = logging.getLogger(__name__)


@dataclass
class PaymentRequest:
    """Payment transaction request"""
    payment_id: str
    workflow_id: str
    from_account_id: bytes  # 16-byte account ID
    to_account_id: bytes    # 16-byte account ID
    amount: int             # Amount in smallest currency unit (e.g., kobo)
    currency: str
    description: str
    tenant_id: str
    user_id: str
    idempotency_key: str


@dataclass
class PaymentResult:
    """Payment transaction result"""
    payment_id: str
    transfer_id: Optional[bytes]
    status: str  # "completed", "failed", "pending"
    timestamp: datetime
    from_account_id: bytes
    to_account_id: bytes
    amount: int
    error: Optional[str] = None


class PaymentProcessor:
    """
    Payment processor with TigerBeetle and Redis integration
    Provides the same functionality as the Go implementation
    """

    def __init__(self, middleware_manager):
        """
        Initialize payment processor
        
        Args:
            middleware_manager: MiddlewareManager instance with Redis, TigerBeetle, Kafka clients
        """
        self.middleware = middleware_manager

    def process_payment_with_locking(
        self,
        req: PaymentRequest,
    ) -> PaymentResult:
        """
        Process a payment transaction with distributed locking
        
        This is the complete implementation showing TigerBeetle and Redis interaction,
        equivalent to the Go ProcessPaymentWithLocking() method.
        
        Steps:
        1. Validate payment request
        2. Check idempotency (Redis)
        3. Acquire distributed lock (Redis)
        4. Cache pending state (Redis)
        5. Publish payment.initiated event (Kafka)
        6. Validate account balances (optional)
        7. Create transfer (TigerBeetle)
        8. Update cache with result (Redis)
        9. Store idempotency key (Redis)
        10. Publish completion event (Kafka)
        11. Release lock (Redis)
        
        Args:
            req: PaymentRequest with payment details
            
        Returns:
            PaymentResult with transaction outcome
            
        Raises:
            ValueError: If request validation fails
            RuntimeError: If lock acquisition fails or transfer fails
        """
        logger.info(
            f"Starting payment processing: {req.payment_id}, "
            f"Amount: {req.amount}"
        )

        # Step 1: Validate payment request
        self._validate_payment_request(req)

        # Step 2: Check for duplicate payment using idempotency key
        idempotency_key = f"payment:idempotency:{req.idempotency_key}"
        existing_result = self.middleware.redis.get_workflow_state(idempotency_key)
        
        if existing_result is not None:
            logger.info(f"Payment already processed (idempotent): {req.payment_id}")
            return PaymentResult(
                payment_id=req.payment_id,
                transfer_id=None,
                status="completed",
                timestamp=datetime.utcnow(),
                from_account_id=req.from_account_id,
                to_account_id=req.to_account_id,
                amount=req.amount,
            )

        # Step 3: Acquire distributed lock to prevent concurrent processing
        lock_name = f"payment:lock:{req.payment_id}"
        lock_timeout = 30  # seconds
        
        logger.info(f"Acquiring distributed lock: {lock_name}")
        locked = self.middleware.redis.acquire_lock(lock_name, lock_timeout)
        
        if not locked:
            logger.warning(f"Payment already being processed: {req.payment_id}")
            raise RuntimeError(f"Payment {req.payment_id} is already being processed")
        
        logger.info(f"Distributed lock acquired: {lock_name}")

        try:
            # Step 4: Cache payment state as "pending" in Redis
            pending_state = {
                "payment_id": req.payment_id,
                "workflow_id": req.workflow_id,
                "status": "pending",
                "from_account": req.from_account_id.hex(),
                "to_account": req.to_account_id.hex(),
                "amount": req.amount,
                "currency": req.currency,
                "timestamp": int(time.time()),
            }
            
            state_key = f"payment:state:{req.payment_id}"
            self.middleware.redis.cache_workflow_state(state_key, pending_state, ttl=3600)
            logger.info(f"Cached pending payment state: {req.payment_id}")

            # Step 5: Publish payment.initiated event to Kafka
            from middleware.kafka.client import WorkflowEvent
            
            initiated_event = WorkflowEvent(
                event_id=f"evt-{req.payment_id}-initiated",
                event_type="payment.initiated",
                workflow_id=req.workflow_id,
                workflow_type="payment",
                status="pending",
                tenant_id=req.tenant_id,
                user_id=req.user_id,
                data={
                    "payment_id": req.payment_id,
                    "amount": req.amount,
                    "currency": req.currency,
                    "from_account": req.from_account_id.hex(),
                    "to_account": req.to_account_id.hex(),
                },
                timestamp=datetime.utcnow(),
            )
            
            try:
                self.middleware.publish_workflow_event(initiated_event)
                logger.info(f"Published payment.initiated event: {req.payment_id}")
            except Exception as e:
                logger.error(f"Failed to publish initiated event: {e}")
                # Continue processing even if event publishing fails

            # Step 6: Validate account balances (optional pre-validation)
            logger.info(f"Validating account balances for: {req.from_account_id.hex()}")
            # This could query TigerBeetle for current balances
            # For now, we proceed directly to transfer creation

            # Step 7: Create transfer in TigerBeetle
            logger.info(
                f"Creating transfer in TigerBeetle: {req.payment_id}, "
                f"Amount: {req.amount}"
            )
            
            # Generate transfer ID from payment ID
            transfer_id = self._generate_transfer_id(req.payment_id)
            
            # Create the transfer using TigerBeetle
            try:
                self.middleware.tigerbeetle.process_payment(
                    payment_id=req.payment_id,
                    from_account_id=req.from_account_id,
                    to_account_id=req.to_account_id,
                    amount=req.amount,
                )
                
                # Transfer succeeded
                logger.info(
                    f"TigerBeetle transfer completed successfully: {req.payment_id}"
                )
                
                result = PaymentResult(
                    payment_id=req.payment_id,
                    transfer_id=transfer_id,
                    status="completed",
                    timestamp=datetime.utcnow(),
                    from_account_id=req.from_account_id,
                    to_account_id=req.to_account_id,
                    amount=req.amount,
                )
                
                # Step 8: Update cache with completed status
                completed_state = {
                    "payment_id": req.payment_id,
                    "transfer_id": transfer_id.hex(),
                    "status": "completed",
                    "from_account": req.from_account_id.hex(),
                    "to_account": req.to_account_id.hex(),
                    "amount": req.amount,
                    "currency": req.currency,
                    "timestamp": int(time.time()),
                }
                
                self.middleware.redis.cache_workflow_state(
                    state_key, completed_state, ttl=3600
                )
                logger.info(f"Updated cache with completed status: {req.payment_id}")
                
                # Step 9: Store idempotency key to prevent duplicate processing
                self.middleware.redis.cache_workflow_state(
                    idempotency_key, completed_state, ttl=86400  # 24 hours
                )
                logger.info(f"Stored idempotency key: {req.payment_id}")
                
                # Step 10: Publish payment.completed event to Kafka
                completed_event = WorkflowEvent(
                    event_id=f"evt-{req.payment_id}-completed",
                    event_type="payment.completed",
                    workflow_id=req.workflow_id,
                    workflow_type="payment",
                    status="completed",
                    tenant_id=req.tenant_id,
                    user_id=req.user_id,
                    data={
                        "payment_id": req.payment_id,
                        "transfer_id": transfer_id.hex(),
                        "amount": req.amount,
                        "currency": req.currency,
                        "from_account": req.from_account_id.hex(),
                        "to_account": req.to_account_id.hex(),
                    },
                    timestamp=datetime.utcnow(),
                )
                
                try:
                    self.middleware.publish_workflow_event(completed_event)
                    logger.info(f"Published payment.completed event: {req.payment_id}")
                except Exception as e:
                    logger.error(f"Failed to publish completed event: {e}")
                    # Don't fail the payment if event publishing fails
                
                logger.info(
                    f"Payment processing completed successfully: {req.payment_id}, "
                    f"Status: {result.status}"
                )
                
                return result
                
            except Exception as e:
                # Transfer failed
                logger.error(f"TigerBeetle transfer failed: {e}")
                
                result = PaymentResult(
                    payment_id=req.payment_id,
                    transfer_id=None,
                    status="failed",
                    timestamp=datetime.utcnow(),
                    from_account_id=req.from_account_id,
                    to_account_id=req.to_account_id,
                    amount=req.amount,
                    error=str(e),
                )
                
                # Update cache with failed status
                failed_state = {
                    "payment_id": req.payment_id,
                    "status": "failed",
                    "error": str(e),
                    "timestamp": int(time.time()),
                }
                
                self.middleware.redis.cache_workflow_state(
                    state_key, failed_state, ttl=3600
                )
                logger.info(f"Updated cache with failed status: {req.payment_id}")
                
                # Publish payment.failed event
                failed_event = WorkflowEvent(
                    event_id=f"evt-{req.payment_id}-failed",
                    event_type="payment.failed",
                    workflow_id=req.workflow_id,
                    workflow_type="payment",
                    status="failed",
                    tenant_id=req.tenant_id,
                    user_id=req.user_id,
                    data={
                        "payment_id": req.payment_id,
                        "amount": req.amount,
                        "error": str(e),
                    },
                    timestamp=datetime.utcnow(),
                )
                
                try:
                    self.middleware.publish_workflow_event(failed_event)
                    logger.info(f"Published payment.failed event: {req.payment_id}")
                except Exception as pub_err:
                    logger.error(f"Failed to publish failed event: {pub_err}")
                
                raise RuntimeError(f"Payment processing failed: {e}") from e
                
        finally:
            # Step 11: Release distributed lock
            logger.info(f"Releasing distributed lock: {lock_name}")
            try:
                self.middleware.redis.release_lock(lock_name)
                logger.info(f"Distributed lock released: {lock_name}")
            except Exception as e:
                logger.error(f"Failed to release lock: {e}")

    def process_payment(
        self,
        payment_id: str,
        from_account_id: bytes,
        to_account_id: bytes,
        amount: int,
    ) -> None:
        """
        Simplified payment processing method (equivalent to Go ProcessPayment)
        
        Args:
            payment_id: Unique payment identifier
            from_account_id: Source account (16 bytes)
            to_account_id: Destination account (16 bytes)
            amount: Amount in smallest currency unit
            
        Raises:
            RuntimeError: If payment processing fails
        """
        req = PaymentRequest(
            payment_id=payment_id,
            workflow_id=payment_id,
            from_account_id=from_account_id,
            to_account_id=to_account_id,
            amount=amount,
            currency="NGN",
            description="Payment transaction",
            tenant_id="",
            user_id="",
            idempotency_key=payment_id,
        )
        
        result = self.process_payment_with_locking(req)
        
        if result.status != "completed":
            raise RuntimeError(f"Payment failed with status: {result.status}")

    def get_payment_status(self, payment_id: str) -> Optional[Dict[str, Any]]:
        """
        Get the current status of a payment from Redis cache
        
        Args:
            payment_id: Payment identifier
            
        Returns:
            Payment state dictionary or None if not found
        """
        state_key = f"payment:state:{payment_id}"
        return self.middleware.redis.get_workflow_state(state_key)

    def cancel_pending_payment(self, payment_id: str) -> None:
        """
        Attempt to cancel a pending payment
        
        Args:
            payment_id: Payment identifier
            
        Raises:
            RuntimeError: If cancellation fails
        """
        # Acquire lock
        lock_name = f"payment:lock:{payment_id}"
        locked = self.middleware.redis.acquire_lock(lock_name, 30)
        
        if not locked:
            raise RuntimeError("Failed to acquire lock for cancellation")
        
        try:
            # Check current status
            state_key = f"payment:state:{payment_id}"
            state = self.middleware.redis.get_workflow_state(state_key)
            
            if state is None:
                raise RuntimeError(f"Payment not found: {payment_id}")
            
            status = state.get("status")
            if status != "pending":
                raise RuntimeError(f"Payment cannot be cancelled (status: {status})")
            
            # Update status to cancelled
            state["status"] = "cancelled"
            state["cancelled_at"] = int(time.time())
            
            self.middleware.redis.cache_workflow_state(state_key, state, ttl=3600)
            logger.info(f"Payment cancelled: {payment_id}")
            
        finally:
            self.middleware.redis.release_lock(lock_name)

    def _validate_payment_request(self, req: PaymentRequest) -> None:
        """
        Validate payment request
        
        Args:
            req: PaymentRequest to validate
            
        Raises:
            ValueError: If validation fails
        """
        if not req.payment_id:
            raise ValueError("payment_id is required")
        
        if req.amount <= 0:
            raise ValueError("amount must be greater than 0")
        
        if req.from_account_id == req.to_account_id:
            raise ValueError("from_account and to_account must be different")
        
        if not req.idempotency_key:
            raise ValueError("idempotency_key is required")
        
        if len(req.from_account_id) != 16:
            raise ValueError("from_account_id must be 16 bytes")
        
        if len(req.to_account_id) != 16:
            raise ValueError("to_account_id must be 16 bytes")

    def _generate_transfer_id(self, payment_id: str) -> bytes:
        """
        Generate a 16-byte transfer ID from payment ID
        
        Args:
            payment_id: Payment identifier string
            
        Returns:
            16-byte transfer ID
        """
        # Convert payment_id to bytes and pad/truncate to 16 bytes
        payment_bytes = payment_id.encode('utf-8')[:16]
        return payment_bytes.ljust(16, b'\x00')


# Example usage
if __name__ == "__main__":
    import logging
    from integration.middleware_manager import MiddlewareManager, MiddlewareConfig
    from middleware.kafka.client import KafkaConfig
    from middleware.redis.client import RedisConfig
    from middleware.tigerbeetle.client import TigerBeetleConfig
    
    # Configure logging
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
    )
    
    # Create middleware configuration
    config = MiddlewareConfig(
        kafka=KafkaConfig(brokers=["localhost:9092"]),
        redis=RedisConfig(),
        tigerbeetle=TigerBeetleConfig(cluster_id=1, addresses=["localhost:3000"]),
        # ... other configs
    )
    
    # Create middleware manager
    middleware = MiddlewareManager(config)
    
    # Create payment processor
    processor = PaymentProcessor(middleware)
    
    # Example: Process a payment
    try:
        # Create account IDs (16 bytes each)
        customer_account = b'CUST-001\x00\x00\x00\x00\x00\x00\x00\x00'
        merchant_account = b'MERCH-001\x00\x00\x00\x00\x00\x00\x00'
        
        # Create payment request
        req = PaymentRequest(
            payment_id="PAY-12345",
            workflow_id="WF-12345",
            from_account_id=customer_account,
            to_account_id=merchant_account,
            amount=50000,  # 500.00 NGN (in kobo)
            currency="NGN",
            description="Product purchase",
            tenant_id="tenant-1",
            user_id="user-123",
            idempotency_key="idempotency-12345",
        )
        
        # Process payment
        result = processor.process_payment_with_locking(req)
        
        print(f"Payment completed: {result.payment_id}")
        print(f"Status: {result.status}")
        print(f"Transfer ID: {result.transfer_id.hex() if result.transfer_id else 'N/A'}")
        print(f"Amount: {result.amount}")
        
        # Check payment status
        status = processor.get_payment_status("PAY-12345")
        print(f"Payment status: {status}")
        
    except Exception as e:
        print(f"Payment failed: {e}")
    
    finally:
        middleware.close()

